### PR TITLE
Open loop sync/async multithreaded testing

### PR DESCRIPTION
**Pull request by vjpai**
This PR adds the ability to do open-loop testing with various random distributions (though only Poisson makes a lot of sense). You can specify the rate as a load parameter, and the client config proto is expanded accordingly.

Note that using open-loop testing effectively will typically require you to set a greater value for outstanding_rpcs_per_channel than you would with closed loop, as you need some slack to make up for times when the channels aren't able to keep up with natural variations in the random distribution.

What's not supported: benchmarking for open-loop async streaming. This is an uncommonly used method of communication, so there wasn't a strong desire to incorporate the additional complexity required to benchmark it.

### CODE REVIEW

**Comment by vjpai** [comment]
Hey, that turned green fairly rapidly. Please review at your convenience.

**Comment by vjpai** [comment]
I've asked @dgquintas for a review since he had been reviewing my earlier failed PR before I closed it.
Thanks!

File: test/cpp/qps/interarrival.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+namespace testing {
+
+// First create classes that define a random distribution
+// Note that this code does not include C++-specific random distribution
+// features supported in std::random. Although this would make this code easier,
+// this code is required to serve as the template code for other language
+// stacks. Thus, this code only uses a uniform distribution of doubles [0,1)
+// and then provides the distribution functions itself.
+
+class RandomDist {
```
Comment (ctiller) [inline test/cpp/qps/interarrival.h]: These distributions need a few lines of documentation on what the distribution looks like.
I'd even take links to Wikipedia articles :)

File: test/cpp/qps/interarrival.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+typedef std::default_random_engine qps_random_engine;
+
+class InterarrivalTimer {
+ public:
+  InterarrivalTimer() {}
+  InterarrivalTimer(const RandomDist& r, int threads, int entries = 1000000) {
+    init(r, threads, entries);
+  }
+  void init(const RandomDist& r, int threads, int entries = 1000000) {
```
Comment (ctiller) [inline test/cpp/qps/interarrival.h]: Why do we need a separate init() function?

File: test/cpp/qps/client_async.cc [hunk 3]
```diff
@@ -159,30 +206,126 @@ class AsyncClient : public Client {
...
+        }
+        // delete the old version
+        delete ctx;
+      }
+      if (!closed_loop_)
+        issue_allowed_[thread_idx] =
+            true;  // may be ok now even if it hadn't been
+    }
+    if (!closed_loop_ && issue_allowed_[thread_idx] &&
+        grpc_time_source::now() >= next_issue_[thread_idx]) {
```
Comment (ctiller) [inline test/cpp/qps/client_async.cc]: The complexity of this versus the sync version makes me think we're missing an abstraction in the API.

Would an alarm that's composable within a completion queue ease this implementation?

File: test/cpp/qps/interarrival.h (hunk 1 above)
Comment (vjpai) [inline test/cpp/qps/interarrival.h]: Will do.

File: test/cpp/qps/interarrival.h (hunk 2 above)
Comment (vjpai) [inline test/cpp/qps/interarrival.h]: Oops - I should deprecate this. This was a planned feature at some point to be able to resize or restart the distribution (for example, to dynamically change the thread count), but I never ended up using it. I will pull it.

File: test/cpp/qps/client_async.cc (hunk 3 above)
Comment (vjpai) [inline test/cpp/qps/client_async.cc]: Yes, definitely. We can talk about this. It would certainly have been nice to insert alarms to indicate rpc-specific timeouts and for knowing when to next issue, rather than having that big mess of checks at the beginning to perfectly set the AsyncNext parameter.

Comment (vjpai) [inline test/cpp/qps/client_async.cc]: One other feature that might reduce complexity and improve performance in the common case would be to actually have separate per-thread per-channel state. Right now there's separate per-thread state and separate per-channel state, but nothing that is per-thread per-channel. Doing that might make sense for the idle contexts structure to avoid locking. The idea right now is that each thread has a preferred subset of the channels but will attempt to use others if those are full. An alternative would be to have each thread freely use each channel (though that just moves the possible mutex contention a layer lower). But this was all the big source of complexity in getting this debugged. Synchronization is hard.

**Comment by ctiller** [comment]
What if we had something like:

```c
namespace grpc {
class Alarm {
 public:
  // emits a completion to cq with tag 'tag' on completion
  void Start(CompletionQueue* cq, void* tag, some_time_type deadline);
  void Cancel();
};
}
```

We'd need a C api to mirror it (grpc_cq_alarm_XXX?), and then we could leverage the existing alarm list implementation to drive it.

**Comment by vjpai** [comment]
That suggestion sounds helpful.

File: include/grpc++/time.h [hunk 4]
```diff
@@ -85,6 +85,9 @@ namespace grpc {
 // from and to should be absolute time.
 void Timepoint2Timespec(const std::chrono::system_clock::time_point& from,
                         gpr_timespec* to);
+void TimepointHR2Timespec(
+    const std::chrono::high_resolution_clock::time_point& from,
+    gpr_timespec* to);
```
Comment (dgquintas) [inline include/grpc++/time.h]: the formatting of these three lines looks funny. Is this how clang-format does it?

File: src/cpp/util/time.cc [hunk 5]
```diff
@@ -59,6 +60,20 @@ void Timepoint2Timespec(const system_clock::time_point& from,
   to->tv_nsec = nsecs.count();
 }
 
+void TimepointHR2Timespec(const high_resolution_clock::time_point& from,
+                          gpr_timespec* to) {
+  high_resolution_clock::duration deadline = from.time_since_epoch();
```
Comment (dgquintas) [inline src/cpp/util/time.cc]: const

File: src/cpp/util/time.cc [hunk 6]
```diff
@@ -59,6 +60,20 @@ void Timepoint2Timespec(const system_clock::time_point& from,
...
+  seconds secs = duration_cast<seconds>(deadline);
```
Comment (dgquintas) [inline src/cpp/util/time.cc]: const

File: src/cpp/util/time.cc [hunk 7]
```diff
@@ -59,6 +60,20 @@ void Timepoint2Timespec(const system_clock::time_point& from,
...
+  if (from == high_resolution_clock::time_point::max() ||
+      secs.count() >= gpr_inf_future.tv_sec || secs.count() < 0) {
+    *to = gpr_inf_future;
+    return;
+  }
+  nanoseconds nsecs = duration_cast<nanoseconds>(deadline - secs);
```
Comment (dgquintas) [inline src/cpp/util/time.cc]: const. I like consts.

File: test/cpp/qps/client.h [hunk 8]
```diff
@@ -42,11 +43,31 @@
...
+
+#if defined(__APPLE__)
+// Specialize Timepoint for high res clock as we need that
+template <>
+class TimePoint<std::chrono::high_resolution_clock::time_point> {
+ public:
+  TimePoint(const std::chrono::high_resolution_clock::time_point& time) {
+    TimepointHR2Timespec(time, &time_);
+  }
+  gpr_timespec raw_time() const { return time_; }
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: super nitpick, but usually lower-cased getter methods match the name of the member variable  they read. s/time_/raw_time_/g ?

File: test/cpp/qps/client.h [hunk 9]
```diff
@@ -106,6 +128,60 @@ class Client {
...
 
+  void SetupLoadTest(const ClientConfig& config, size_t num_threads) {
+    // Set up the load distribution based on the number of threads
+    if (config.load_type() == CLOSED_LOOP) {
+      closed_loop_ = true;
+    } else {
+      closed_loop_ = false;
+
+      std::unique_ptr<RandomDist> random_dist;
+      auto& load = config.load_params();
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: Avoid non-const references. We should be able to make this const (it looks like the methods invoked over this instance are or should/could be const). If it needs be mutable, we should use a pointer.

File: test/cpp/qps/client.h [hunk 10]
```diff
@@ -106,6 +128,60 @@ class Client {
...
+      interarrival_timer_.init(*random_dist, num_threads);
+      for (size_t i = 0; i < num_threads; i++) {
+        next_time_.push_back(
+            grpc_time_source::now() +
+            std::chrono::duration_cast<grpc_time_source::duration>(
+                interarrival_timer_(i)));
+      }
+    }
+  }
+  bool NextIssueTime(int thread_idx, grpc_time* time_delay) {
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: nit: separate methods with empty line

File: test/cpp/qps/interarrival.h (hunk 2 above)
Comment (vjpai) [inline test/cpp/qps/interarrival.h]: So, I acted the other way. I removed the parameterized constructor. I kept the init function. The reason is that we don't know right away what the useful parameters are for the random distribution until we've already parsed the proto in Client, not right when we construct Client.

File: test/cpp/qps/interarrival.h (hunk 1 above)
Comment (vjpai) [inline test/cpp/qps/interarrival.h]: Done.

File: include/grpc++/time.h (hunk 4 above)
Comment (vjpai) [inline include/grpc++/time.h]: Yes, this was courtesy of clang-format.

File: src/cpp/util/time.cc (hunk 5 above)
Comment (vjpai) [inline src/cpp/util/time.cc]: What should be const? deadline? Do we care? I essentially copied the existing Timepoint2Timespec code.

File: src/cpp/util/time.cc (hunk 7 above)
Comment (vjpai) [inline src/cpp/util/time.cc]: This seems like a code health issue separate from this PR since it applies in the function above it as well.

File: test/cpp/qps/client.h (hunk 8 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: Ok, this was also a copy of the specialization for system_clock. Can we do this in a separate PR for consistency?

File: src/cpp/util/time.cc (hunk 7 above)
Comment (dgquintas) [inline src/cpp/util/time.cc]: Yeah, agreed. I feel we don't use const as much as we should.

File: test/cpp/qps/client.h (hunk 8 above)
Comment (dgquintas) [inline test/cpp/qps/client.h]: +1

File: test/cpp/qps/client.h (hunk 9 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: Done!

File: test/cpp/qps/client.h (hunk 10 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: Done

File: test/cpp/qps/client_async.cc [hunk 11]
```diff
@@ -118,29 +139,55 @@ class ClientRpcContextUnaryImpl : public ClientRpcContext {
...
+        next_issue_.push_back(next_issue);
+      }
     }
+
     int t = 0;
     for (int i = 0; i < config.outstanding_rpcs_per_channel(); i++) {
-      for (auto channel = channels_.begin(); channel != channels_.end();
-           channel++) {
+      for (int ch = 0; ch < channel_count_; ch++) {
+        auto& channel = channels_[ch];
```
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: const & or pointer

Comment (vjpai) [inline test/cpp/qps/client_async.cc]: Better, will just directly place the reference into the code below.

**Comment by vjpai** [comment]
Hi there, any more comments, or is this a go at this point?

**Comment by ctiller** [comment]
Can you add a variation of qps_test that sets up and runs a fixed rate test for ten seconds on localhost? It'll let this code get some more miles on Travis.

**Comment by vjpai** [comment]
OK, will do.

**Comment by vjpai** [comment]
So, the new tests are in. The tests themselves pass, but coveralls seems to be stuck.
